File: Multithreaded Programming/fine-grained-bank-ppv-vldmr/java/BankImpl.java
/**
 * Bank implementation.
 *
 * <p>:TODO: This implementation has to be made thread-safe.
 *
 * @author :TODO: Popov Vladimir
 */
public class BankImpl implements Bank {
    /**
     * Private account data structure.
     */
    static class Account {
        /**
         * Amount of funds in this account.
         */
        long amount;
    }

    /**
     * An array of accounts by index.
     */
    private final Account[] accountsOfClients;

    /**
     * Creates new bank instance.
     * @param n the number of accounts (numbered from 0 to n-1).
     */
    public BankImpl(int n) {
        accountsOfClients = new Account[n];
        for (int i = 0; i < n; i++) {
            accountsOfClients[i] = new Account();
        }
    }

    @Override
    public int getNumberOfAccounts() {
        return accountsOfClients.length;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long getAmount(int index) {
        return accountsOfClients[index].amount;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long getTotalAmount() {
        long total = 0;
        for (Account accountsOfClient : accountsOfClients) {
            total += accountsOfClient.amount;
        }
        return total;
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long deposit(int index, long amount) {
        if (amount > 0) {
            Account current = accountsOfClients[index];
            if (amount <= MAX_AMOUNT && current.amount + amount <= MAX_AMOUNT) {
                current.amount += amount;
                return current.amount;
            } else {
                throw new IllegalStateException("Overflow");
            }
        } else {
            throw new IllegalArgumentException("Invalid amount: " + amount);
        }
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public long withdraw(int index, long amount) {
        if (amount > 0) {
            Account current = accountsOfClients[index];
            if (current.amount - amount < 0) {
                throw new IllegalStateException("Underflow");
            }
            current.amount -= amount;
            return current.amount;
        } else {
            throw new IllegalArgumentException("Invalid amount: " + amount);
        }
    }

    /**
     * <p>:TODO: This method has to be made thread-safe.
     */
    @Override
    public void transfer(int fromIndex, int toIndex, long amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("Invalid amount: " + amount);
        } else {
            if (fromIndex == toIndex) {
                throw new IllegalArgumentException("fromIndex == toIndex");
            } else {
                Account l = accountsOfClients[fromIndex];
                Account r = accountsOfClients[toIndex];
                if (amount > l.amount) {
                    throw new IllegalStateException("Underflow");
                } else  {
                    if (amount > MAX_AMOUNT || r.amount + amount > MAX_AMOUNT) {
                        throw new IllegalStateException("Overflow");
                    }
                }
                l.amount -= amount;
                r.amount += amount;
            }
        }
    }
}

File: Multithreaded Programming/fine-grained-bank-ppv-vldmr/src/BankImpl.kt
import java.util.concurrent.locks.ReentrantLock


/**
 * Bank implementation.
 *
 *
 * :
 *
 * @author : Popov Vladimir
 */
/**
 * Creates new bank instance.
 *
 * @param n the number of accounts (numbered from 0 to n-1).
 */
class BankImpl (n: Int) : Bank {
    /**
     * An array of accounts by index.
     */
    private val accounts: Array<Account> = Array(n) { Account() }

    override val numberOfAccounts: Int
        get() = accounts.size

    /**
     *
     * :TODO: DONE
     */
    override val totalAmount: Long
        get() {
            try {
                var total: Long = 0
                for (acc in accounts) {
                    acc.lock.lock()
                    total += acc.amount
                }
                return total
            } finally {
                for (acc in accounts) {
                    acc.lock.unlock()
                }
            }
        }

    /**
     *
     * :TODO: DONE
     */
    override fun getAmount(index: Int): Long {
        accounts[index].lock.lock()
        try {
            return accounts[index].amount
        } finally {
            accounts[index].lock.unlock()
        }
    }

    /**
     *
     * :TODO: DONE
     */
    override fun deposit(index: Int, amount: Long): Long {
        require(amount > 0) { "Invalid amount: $amount" }
        val acc = accounts[index]
        acc.lock.lock()
        try {
            check(!(amount > Bank.MAX_AMOUNT || acc.amount + amount > Bank.MAX_AMOUNT)) { "Overflow" }
            acc.amount += amount
            return acc.amount
        } finally {
            acc.lock.unlock()
        }
    }

    /**
     *
     * :TODO: DONE
     */
    override fun withdraw(index: Int, amount: Long): Long {
        require(amount > 0) { "Invalid amount: $amount" }
        val acc = accounts[index]
        acc.lock.lock()
        try {
            check(acc.amount - amount >= 0) { "Underflow" }
            acc.amount -= amount
            return acc.amount
        } finally {
            acc.lock.unlock()
        }
    }

    /**
     *
     * :TODO: DONE.
     */
    override fun transfer(fromIndex: Int, toIndex: Int, amount: Long) {
        require(amount > 0) { "Invalid amount: $amount" }
        require(fromIndex != toIndex) { "fromIndex == toIndex" }
        val fromAcc = accounts[fromIndex]
        val toAcc = accounts[toIndex]
        if (fromIndex < toIndex) {
            fromAcc.lock.lock()
            toAcc.lock.lock()
        } else {
            toAcc.lock.lock()
            fromAcc.lock.lock()
        }
        try {
            check(amount <= fromAcc.amount) { "Underflow" }
            check(!(amount > Bank.MAX_AMOUNT || toAcc.amount + amount > Bank.MAX_AMOUNT)) { "Overflow" }
            fromAcc.amount -= amount
            toAcc.amount += amount
        } finally {
            if (fromIndex < toIndex) {
                toAcc.lock.unlock()
                fromAcc.lock.unlock()
            } else {
                fromAcc.lock.unlock()
                toAcc.lock.unlock()
            }
        }
    }

    /**
     * Private account data structure.
     */
    internal class Account {
        /**
         * Amount of funds in this account.
         */
        var lock = ReentrantLock()
        var amount: Long = 0
    }
}